### ISSUE TITLE
 VBCompiler.CompileAssemblyFromFileBatch() is modifying the passed in CompilerOptions

### ISSUE THREAD

**Issue by niemyjski** [aspnet/RoslynCodeDomProvider#8]
I was using the legacy VB CodeDom provider to generate VB Code and compile it on the fly for our app. It was working great but we wanted to support newer language features so we brought in the latest Roslyn Code Dom provider via nuget... The change was simple and our c# implementation works flawlessly but I can't even compile the simplest sample due to the VBCompiler is injecting unwanted/invalid compiler options

Lets say I pass in the compiler options of `/optionInfer+ /define:CodeSmith40 /define:CodeSmith50 /define:CodeSmith60 /define:CodeSmith70 /define:CodeSmith80 /define:TRACE` they are modified once I call CompileAssemblyFromFileBatch to `/nowarn:41008,40000,40008 /define:_MYTYPE=\\\"Web\\\" /imports:Microsoft.VisualBasic  /optionInfer+ /define:CodeSmith40 /define:CodeSmith50 /define:CodeSmith60 /define:CodeSmith70 /define:CodeSmith80 /define:TRACE`

Once it defines `_MYTYPE` (unwanted), I believe it's calling the following code (https://github.com/dotnet/roslyn/blob/614299ff83da9959fa07131c6d0ffbc58873b6ae/src/Compilers/VisualBasic/Portable/Symbols/EmbeddedSymbols/VbMyTemplateText.vb#L287) which causes my compile to fail with:

```
vbc : (1,1): error BC30002: Type 'Global.System.Web.HttpRequest' is not defined.
vbc : (1,1): error BC30002: Type 'Global.System.Web.HttpResponse' is not defined.
```

```
  <package id="Microsoft.CodeDom.Providers.DotNetCompilerPlatform" version="1.0.5" targetFramework="net462" />
  <package id="Microsoft.Net.Compilers" version="2.2.0" targetFramework="net462" developmentDependency="true" />
```

https://github.com/dotnet/roslyn/issues/20214

**Comment by Jinhuafei** [comment]
@jaredpar /define:_MYTYPE="Web" has been there for years, and it's been working with legacy vbc.exe without any issue and @niemyjski didn't see any issue with old one. It looks like the issue in new vbc.exe.

**Comment by Jinhuafei** [comment]
@niemyjski would you please share some code snippet to repro the issue?

**Comment by niemyjski** [comment]
It's kind of complicated to break down down from what we do. Basically the process is simple, we build up a list of references, source files (created from a compilation unit (code pasted below) and compiler options and we build up the compiler context and then call the method above to generate an assembly from the generated source code from the previous step (we've been doing this for over a decade and it's always worked.

```cs
// Generate the source file that's below the images
 if (context.Request.CodeCompileUnit != null)
            {
                string codeDomSourceFile = Path.Combine(context.TempFiles.TempDir, String.Format("{0}.{1}", context.AssemblyName, context.Provider.FileExtension));
                using (var writer = new StreamWriter(codeDomSourceFile))
                {
                    ICodeGenerator codeGenerator = context.Provider.CreateGenerator();
                    var codeGeneratorOptions = new CodeGeneratorOptions();
                    codeGenerator.GenerateCodeFromCompileUnit(context.Request.CodeCompileUnit, writer, codeGeneratorOptions);
                }

                context.SourceFiles.Add(codeDomSourceFile);
                context.TempFiles.AddFile(codeDomSourceFile, context.TempFiles.KeepFiles);
```

<img width="1071" alt="screen shot 2017-06-16 at 3 52 37 pm" src="https://user-images.githubusercontent.com/1020579/27244460-38882582-52ac-11e7-9678-d7bb158566e1.png">
<img width="1142" alt="screen shot 2017-06-16 at 3 54 01 pm" src="https://user-images.githubusercontent.com/1020579/27244461-3896b732-52ac-11e7-9f29-f2b996ac699e.png">

```vb
'------------------------------------------------------------------------------
' <auto-generated>
'     This code was generated by a tool.
'     Runtime Version:4.0.30319.42000
'
'     Changes to this file may cause incorrect behavior and will be lost if
'     the code is regenerated.
' </auto-generated>
'------------------------------------------------------------------------------

Option Strict Off
Option Explicit On

Imports CodeSmith.Engine
Imports Microsoft.VisualBasic
Imports System
Imports System.ComponentModel
Imports System.Data
Imports System.Diagnostics
Imports System.Linq

Namespace _CodeSmith
    
    Partial Public Class Untitled1_cst
        Inherits CodeSmith.Engine.CodeTemplate
        
        Public Sub New()
            MyBase.New()
        End Sub
        
        <Browsable(false)>  _
        Public Overrides ReadOnly Property CodeTemplateInfo() As CodeSmith.Engine.ICodeTemplateInfo
            Get
                Return New RuntimeCodeTemplateInfo()
            End Get
        End Property
        
        Protected Overrides Sub CreateChildControls()
            Dim delegateControl1 As CodeSmith.Engine.DelegateControl = New CodeSmith.Engine.DelegateControl()
            delegateControl1.RenderMethod = AddressOf Me.__RenderMethod1
            Me.Controls.Add(delegateControl1)
        End Sub
        
        Protected Overrides Sub SetPropertyDefaults()
        End Sub
        
        Protected Overrides Sub SetUnknownDirectiveAttributes()
        End Sub
        
        Private Sub __RenderMethod1(ByVal writer As System.IO.TextWriter, ByVal control As CodeSmith.Engine.Control)
            writer.Write("test")
        End Sub
        
        Public Overrides Function GetHashCode() As Integer
            Return 62395584
        End Function
        
        Public Class RuntimeCodeTemplateInfo
            Inherits Object
            Implements CodeSmith.Engine.ICodeTemplateInfo
            
            Friend Sub New()
                MyBase.New
            End Sub
            
            Public ReadOnly Property DateCreated() As Date Implements CodeSmith.Engine.ICodeTemplateInfo.DateCreated
                Get
                    Return New Date(504911016000000000)
                End Get
            End Property
            
            Public ReadOnly Property DateModified() As Date Implements CodeSmith.Engine.ICodeTemplateInfo.DateModified
                Get
                    Return New Date(504911016000000000)
                End Get
            End Property
            
            Public ReadOnly Property Description() As String Implements CodeSmith.Engine.ICodeTemplateInfo.Description
                Get
                    Return ""
                End Get
            End Property
            
            Public ReadOnly Property DirectoryName() As String Implements CodeSmith.Engine.ICodeTemplateInfo.DirectoryName
                Get
                    Return "C:\Users\blake\AppData\Local\Temp\"
                End Get
            End Property
            
            Public ReadOnly Property FileName() As String Implements CodeSmith.Engine.ICodeTemplateInfo.FileName
                Get
                    Return "Untitled1.cst"
                End Get
            End Property
            
            Public ReadOnly Property FullPath() As String Implements CodeSmith.Engine.ICodeTemplateInfo.FullPath
                Get
                    Return "C:\Users\blake\AppData\Local\Temp\Untitled1.cst"
                End Get
            End Property
            
            Public ReadOnly Property CodeBehind() As String Implements CodeSmith.Engine.ICodeTemplateInfo.CodeBehind
                Get
                    Return ""
                End Get
            End Property
            
            Public ReadOnly Property Language() As String Implements CodeSmith.Engine.ICodeTemplateInfo.Language
                Get
                    Return "VB"
                End Get
            End Property
            
            Public ReadOnly Property TargetLanguage() As String Implements CodeSmith.Engine.ICodeTemplateInfo.TargetLanguage
                Get
                    Return ""
                End Get
            End Property
            
            Public ReadOnly Property TemplateEncoding() As String Implements CodeSmith.Engine.ICodeTemplateInfo.TemplateEncoding
                Get
                    Return ""
                End Get
            End Property
            
            Public ReadOnly Property OutputType() As CodeSmith.Engine.CodeTemplateOutputType Implements CodeSmith.Engine.ICodeTemplateInfo.OutputType
                Get
                    Return CodeTemplateOutputType.Normal
                End Get
            End Property
            
            Public ReadOnly Property CodeSmithVersion() As String Implements CodeSmith.Engine.ICodeTemplateInfo.CodeSmithVersion
                Get
                    Return "8.0.14319"
                End Get
            End Property
            
            Public ReadOnly Property FileDependencies() As String() Implements CodeSmith.Engine.ICodeTemplateInfo.FileDependencies
                Get
                    Return New String() {"C:\Users\blake\AppData\Local\Temp\Untitled1.cst"}
                End Get
            End Property
            
            Public ReadOnly Property AssemblyDependencies() As String() Implements CodeSmith.Engine.ICodeTemplateInfo.AssemblyDependencies
                Get
                    Return New String(-1) {}
                End Get
            End Property
            
            Public Function GetContent() As String Implements CodeSmith.Engine.ICodeTemplateInfo.GetContent
                Return String.Empty
            End Function
            
            Public Sub Refresh() Implements CodeSmith.Engine.ICodeTemplateInfo.Refresh
            End Sub
            
            Public Overrides Function GetHashCode() As Integer
                Return 62395584
            End Function
        End Class
    End Class
End Namespace
```

As you can see we don't specify any other references other than those shown in the images above and our code has no references to system.web.

**Comment by Jinhuafei** [comment]
@niemyjski  _MYTYPE is added to support [this feature](https://docs.microsoft.com/en-us/dotnet/visual-basic/developing-apps/customizing-extending-my/customizing-which-objects-are-available-in-my), and compiler does some tricks to support it. RoslynCodeDomProvider is mainly for compiling the code dynamically generated by asp.net, that's why we set _MYTYPE to "web".  

I used old vb.net compiler to compile the vb code you provided with _MYTYPE="Web" option. I got the similar error too if I don't reference system.web.dll. You mentioned your code worked before, so you may want to check if you added reference to system.web.dll.

**Comment by niemyjski** [comment]
We are not adding a reference to system.web and have no plans to. We are using this in a desktop app for code generation. It would be great if this was made generic and didn't make any assumptions like a generic nuget package required to reference system.web and was actually owned by the .net team.

**Comment by Jinhuafei** [comment]
We will consider your suggestion.

**Comment by niemyjski** [comment]
How can you consider it if the issue is closed?

**Comment by Jinhuafei** [comment]
I added enhancement tag and clicked the wrong button. :)

**Comment by dlumpp** [comment]
I don't know if there are plans to address this but I just want to add that I lost a lot of time troubleshooting a 3rd party closed source library that was broken only when referenced in an MVC project because of this issue. I would have never figured it out if this issue hadn't come up in my searching. 

I agree with @niemyjski , it doesn't seem right to expect every library to take a dependency on System.Web in order use the Roslyn compiler. 

I had to delete the VB provider entirely from web.config to solve my problem. Removing the _MYTYPE Web declaration didn't work.

**Comment by WolfgangHG** [comment]
Had the same issue when trying to migrate an old CodeDomProvider app to roslyn (WinForms application, no web assemblies references), and found this issue.

My workaround was to add "define _MYTYPE=Custom" to the compiler parameters:

`compilerParameters.CompilerOptions = $"/optionexplicit+ /optionstrict+ /debug- /debug:pdbonly /define:_MYTYPE=\\\"Custom\\\"";`

It seems to work after doing this.

**Comment by StephenMolloy** [comment]
I believe this should be fixed in 3.5 with #102. We will still muddle the compiler options when running in ASP.Net by default, but using the codedom providers directly should turn that off. Or you can explicitly decide to turn this ASP.Net magic stuff on or off with a provider option. (See the updated Readme.md for details.)
